File: Freedom/微信/基本类库/Common/WXWebViewController.m
//  TLWebViewController.m
//  Freedom
// Created by Super
#import "WXWebViewController.h"
#import <UMMobClick/MobClick.h>
#define     WEBVIEW_NAVBAR_ITEMS_FIXED_SPACE    9
@interface WXWebViewController ()
@property (nonatomic, strong) WKWebView *webView;
@property (nonatomic, strong) UIProgressView *progressView;
@property (nonatomic, strong) UIBarButtonItem *backButtonItem;
@property (nonatomic, strong) UIBarButtonItem *closeButtonItem;
@property (nonatomic, strong) UILabel *authLabel;
@end
@implementation WXWebViewController
- (id)init{
    if (self = [super init]) {
        self.useMPageTitleAsNavTitle = YES;
        self.showLoadingProgress = YES;
    }
    return self;
}
- (void)loadView{
    [super loadView];
    [self.view addSubview:self.authLabel];
    [self.view addSubview:self.webView];
    [self.view addSubview:self.progressView];
}
- (void)viewDidLoad{
    [super viewDidLoad];
    [self setAutomaticallyAdjustsScrollViewInsets:NO];
    [self.view setBackgroundColor:UIColor(46.0, 49.0, 50.0, 1.0)];
    [self.webView.scrollView setBackgroundColor:[UIColor clearColor]];
    for (id vc in self.webView.scrollView.subviews) {
        NSString *className = NSStringFromClass([vc class]);
        if ([className isEqualToString:@"WKContentView"]) {
            [vc setBackgroundColor:[UIColor whiteColor]];
            break;
        }
    }
    [self.webView addObserver:self forKeyPath:@"estimatedProgress" options:NSKeyValueObservingOptionNew context:NULL];
    [self.webView.scrollView addObserver:self forKeyPath:@"backgroundColor" options:NSKeyValueObservingOptionNew context:NULL];
}
- (void)viewWillAppear:(BOOL)animated{
    [super viewWillAppear:animated];
    [MobClick beginLogPageView:@"WebVC"];
    [self.progressView setProgress:0.0f];
    [self.webView loadRequest:[NSURLRequest requestWithURL:[NSURL URLWithString:self.url]]];
    if (!self.disableBackButton && self.navigationItem.leftBarButtonItems.count <= 2) {
        [self.navigationItem setLeftBarButtonItems:@[self.backButtonItem]];
    }
}
- (void)viewWillDisappear:(BOOL)animated{
    [super viewWillDisappear:animated];
    [MobClick endLogPageView:@"WebVC"];
}
- (void)dealloc{
    [self.webView removeObserver:self forKeyPath:@"estimatedProgress"];
    [self.webView.scrollView removeObserver:self forKeyPath:@"backgroundColor"];
#ifdef DEBUG_MEMERY
    DLog(@"dealloc WebVC");
#endif
}
#pragma mark - Public Methods -
- (void)setUrl:(NSString *)url{
    _url = url;
    if ([self.view isFirstResponder]) {
        [self.progressView setProgress:0.0f];
        [self.webView loadRequest:[NSURLRequest requestWithURL:[NSURL URLWithString:self.url]]];
    }
}
- (void)setShowLoadingProgress:(BOOL)showLoadingProgress{
    _showLoadingProgress = showLoadingProgress;
    [self.progressView setHidden:!showLoadingProgress];
}
#pragma mark - Event Response -
- (void)observeValueForKeyPath:(NSString *)keyPath ofObject:(id)object change:(NSDictionary *)change context:(void *)context {
    
    if (self.showLoadingProgress && [keyPath isEqualToString:@"estimatedProgress"] && object == self.webView) {
        [self.progressView setAlpha:1.0f];
        [self.progressView setProgress:self.webView.estimatedProgress animated:YES];
        
        if(self.webView.estimatedProgress >= 1.0f) {
            [UIView animateWithDuration:0.3 delay:0.3 options:UIViewAnimationOptionCurveEaseOut animations:^{
                [self.progressView setAlpha:0.0f];
            } completion:^(BOOL finished) {
                [self.progressView setProgress:0.0f animated:NO];
            }];
        }
    }else if ([keyPath isEqualToString:@"backgroundColor"] && object == self.webView.scrollView) {
        UIColor *color = [change objectForKey:@"new"];
        if (!CGColorEqualToColor(color.CGColor, [UIColor clearColor].CGColor)) {
            [object setBackgroundColor:[UIColor clearColor]];
        }
    }
}
- (void)navBackButotnDown{
    if (self.webView.canGoBack) {
        [self.webView goBack];
        [self.navigationItem setLeftBarButtonItems:@[self.backButtonItem, self.closeButtonItem]];
    }else{
        [self.navigationController popViewControllerAnimated:YES];
    }
}
- (void)navCloseButtonDown{
    [self.navigationController popViewControllerAnimated:YES];
}
#pragma mark - Delegate -
//MARK: WKNavigationDelegate
- (void)webView:(WKWebView *)webView didFinishNavigation:(WKNavigation *)navigation{
    if (self.useMPageTitleAsNavTitle) {
        [self.navigationItem setTitle:webView.title];
        [self.authLabel setText:[NSString stringWithFormat:@"网页由 %@ 提供", webView.URL.host]];
        [self.authLabel setFrameHeight:[self.authLabel sizeThatFits:CGSizeMake(self.authLabel.frame.size.width, MAXFLOAT)].height];
    }
}
#pragma mark - Getter -
- (WKWebView *)webView{
    if (_webView == nil) {
        _webView = [[WKWebView alloc] initWithFrame:CGRectMake(0, TopHeight + 20, APPW, APPH - TopHeight - 20)];
        [_webView setAllowsBackForwardNavigationGestures:YES];
        [_webView setNavigationDelegate:self];
    }
    return _webView;
}
- (UIProgressView *)progressView{
    if (_progressView == nil) {
        _progressView = [[UIProgressView alloc] initWithFrame:CGRectMake(0, TopHeight + 20, APPW, 10.0f)];
        [_progressView setTransform: CGAffineTransformMakeScale(1.0f, 2.0f)];
        [_progressView setProgressTintColor:[UIColor greenColor]];
        [_progressView setTrackTintColor:[UIColor clearColor]];
    }
    return _progressView;
}
- (UIBarButtonItem *)backButtonItem{
    if (_backButtonItem == nil) {
        _backButtonItem = [[UIBarButtonItem alloc] initWithBackTitle:@"返回" target:self action:@selector(navBackButotnDown)];
    }
    return _backButtonItem;
}
- (UIBarButtonItem *)closeButtonItem{
    if (_closeButtonItem == nil) {
        _closeButtonItem = [[UIBarButtonItem alloc] initWithTitle:@"关闭" style:UIBarButtonItemStylePlain target:self action:@selector(navCloseButtonDown)];
    }
    return _closeButtonItem;
}
- (UILabel *)authLabel{
    if (_authLabel == nil) {
        _authLabel = [[UILabel alloc] initWithFrame:CGRectMake(20, TopHeight + 20 + 13, APPW - 40, 0)];
        [_authLabel setFont:[UIFont systemFontOfSize:12.0f]];
        [_authLabel setTextAlignment:NSTextAlignmentCenter];
        [_authLabel setTextColor:[UIColor grayColor]];
        [_authLabel setNumberOfLines:0];
    }
    return _authLabel;
}
@end

File: Freedom/微信/基本类库/Common/WXWebViewController.swift
//
//  WXWebViewController.swift
//  Freedom
import Foundation
import WebKit
class WXWebViewController: WXBaseViewController, WKNavigationDelegate {/// 是否使用网页标题作为nav标题，默认YES
    var useMPageTitleAsNavTitle = true/// 是否显示加载进度，默认YES
    // 是否禁止历史记录，默认NO
    var showLoadingProgress = true {
        didSet {
            progressView.isHidden = !showLoadingProgress
        }
    }
    var disableBackButton = false
    var url = "" {
        didSet {
            if view.isFirstResponder {
                progressView.progress = 0.0
                if let anUrl = URL(string: self.url) {
                    webView.load(URLRequest(url: anUrl))
                }
            }
        }
    }
    lazy var webView = WKWebView(frame: CGRect(x: 0, y: Int(TopHeight) + 20, width: Int(APPW), height: Int(APPH - TopHeight) - 20))
    lazy var progressView: UIProgressView = {
        let progressView = UIProgressView(frame: CGRect(x: 0, y: TopHeight + 20.0, width: APPW, height: 10.0))
        progressView.transform = CGAffineTransform(scaleX: 1.0, y: 2.0)
        progressView.progressTintColor = UIColor.green
        progressView.trackTintColor = UIColor.clear
        return progressView
    }()
    lazy var backButtonItem = UIBarButtonItem(title: "返回", style: UIBarButtonItem.Style.done, actionBlick: {
        if self.webView.canGoBack {
            self.webView.goBack()
            self.navigationItem.leftBarButtonItems = [self.backButtonItem, self.closeButtonItem] as? [UIBarButtonItem]
        } else {
            self.navigationController?.popViewController(animated: true)
        }
    })
    lazy var closeButtonItem = UIBarButtonItem(title: "关闭") {
        self.navigationController?.popViewController(animated: true)
    }
    lazy var authLabel: UILabel = {
        let authLabel = UILabel(frame: CGRect(x: 20, y: Int(TopHeight) + 20 + 13, width: Int(APPW - 40), height: 0))
            authLabel.font = UIFont.systemFont(ofSize: 12.0)
            authLabel.textAlignment = .center
            authLabel.textColor = UIColor.gray
            authLabel.numberOfLines = 0
            return authLabel
        }()

    override func loadView() {
        super.loadView()
        webView.allowsBackForwardNavigationGestures = true
        webView.navigationDelegate = self
        view.addSubview(authLabel)
        view.addSubview(webView)
        view.addSubview(progressView)
    }

    override func viewDidLoad() {
        super.viewDidLoad()
        automaticallyAdjustsScrollViewInsets = false
        view.backgroundColor = UIColor(46.0, 49.0, 50.0, 1.0)
        webView.scrollView.backgroundColor = UIColor.clear
        for vc in webView.scrollView.subviews {
            if let cls = object_getClass(vc) {
                if (NSStringFromClass(cls) == "WKContentView") {
                    vc.backgroundColor = UIColor.white
                    break
                }
            }
        }
        webView.addObserver(self, forKeyPath: "estimatedProgress", options: .new, context: nil)
        webView.scrollView.addObserver(self, forKeyPath: "backgroundColor", options: .new, context: nil)
    }

    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        MobClick.beginLogPageView("WebVC")
        progressView.progress = 0.0
        if let anUrl = URL(string: url) {
            webView.load(URLRequest(url: anUrl))
        }
        if !disableBackButton && (navigationItem.leftBarButtonItems?.count ?? 0) <= 2 {
            navigationItem.leftBarButtonItems = [backButtonItem] as? [UIBarButtonItem]
        }
    }
    deinit {
        webView.removeObserver(self, forKeyPath: "estimatedProgress")
        webView.scrollView.removeObserver(self, forKeyPath: "backgroundColor")
    }
    func observeValue(forKeyPath keyPath: String, of object: Any, change: [NSKeyValueChangeKey : Any], context: UnsafeMutableRawPointer) {
        if showLoadingProgress && (keyPath == "estimatedProgress") && (object as! WKWebView) == webView {
            progressView.alpha = 1.0
            progressView.setProgress(Float(webView.estimatedProgress), animated: true)
            if webView.estimatedProgress >= 1.0 {
                UIView.animate(withDuration: 0.3, delay: 0.3, options: .curveEaseOut, animations: {
                    self.progressView.alpha = 0.0
                }) { finished in
                    self.progressView.setProgress(0.0, animated: false)
                }
            }
        } else if (keyPath == "backgroundColor") && (object as! UIScrollView) == webView.scrollView {
            let color = change[NSKeyValueChangeKey.newKey] as! UIColor
            if !(color.cgColor == UIColor.clear.cgColor) {
                (object as! UIView).backgroundColor = UIColor.clear
            }
        }
    }
    func webView(_ webView: WKWebView, didFinish navigation: WKNavigation!) {
        if useMPageTitleAsNavTitle {
            navigationItem.title = webView.title
            authLabel.text = "网页由 \(String(describing: webView.url?.host)) 提供"
            var rec: CGRect = authLabel.frame
            rec.size.height = authLabel.sizeThatFits(CGSize(width: authLabel.frame.size.width, height: CGFloat(MAXFLOAT))).height
            authLabel.frame = rec
        }
    }
}
